### DISCUSSION TITLE
Is it possible to retrieve only from a subset of the corpus?

### DISCUSSION THREAD

**Discussion by tsdev** [xhluca/bm25s#82] [Q&A] [answered]
Is it possible to retrieve only from the subset of the corpus? I would like to index the whole corpus but then only retrieve from a subset of the corpus. E.g. I have 1M sentences as corpus, I index it to get good word frequencies for the BM25 scores. Then I would like to retrieve k=1 but only from the first 1k sentences. Is this possible?

**Comment by xhluca** [comment, marked as answer]
You can use the `weight_mask` parameter in retrieve and use an array of 1 for docs you want to retrieve, and 0 for others.

Here's an example: 

https://github.com/xhluca/bm25s/blob/8b5ff105dc550189412da83be68de9a857b325f5/tests/core/test_retrieve.py#L71-L105

**Comment by TheEpicBigBoss** [comment]
But the `weight_mask` is just multiplied after calculating the scores. So the retrieval is still done on the full corpus. Is there a way to only calculate the scores on a subset of the corpus / index to actually speed up the retrieval?

**Reply by xhluca** [reply]
Not at the moment. It should be possible to implement but could become quite tricky to implement. Do you have an idea how to approach this?

**Reply by TheEpicBigBoss** [reply]
I am not so deep into your code because I am switching to my own implementation, but something like this, where the bm25 scores are only calculated on a subset of the entire matrix, should be also possible in your package:

```python 
# with f being a sparse frequency matrix of each token per document
freq = csr_matrix((N, vocab_size))

# specifying indices for the subset
subset_indices = [11, 22, 33]
subset_freq = freq[subset_indices]
scores = BM25_calc(subset_freq, query)

# zip scores with original indices and sort
results = list(zip(subset_indices, scores))
results.sort(key=lambda x: x[1], reverse=True)
```

This will not reduce the memory overhead but the retrieval speed is mostly the main problem (for me).
